torch/csrc/autograd/record_function.cpp

```diff
@@ -11,177 +12,192 @@ namespace profiler {
+    mergeRunCallbacks(
+        sorted_global_callbacks_,
+        rf.sorted_active_global_handles_,
+        /* is_start */ true,
+        rf);
+    mergeRunCallbacks(
+        sorted_tls_callbacks_,
+        rf.sorted_active_tls_handles_,
+        /* is_start */ true,
+        rf);
   }
 
-  inline double samplingProbability(size_t cb_idx) const {
-    TORCH_INTERNAL_ASSERT(cb_idx < callbacks_.size());
-    if (callbacks_[cb_idx].is_sampled_) {
-      return use_global_prob_ ? global_prob_ : callbacks_[cb_idx].sampling_prob_;
-    } else {
-      return 1.0;
-    }
+  void runEndCallbacks(RecordFunction& rf) {
+    mergeRunCallbacks(
+        sorted_global_callbacks_,
+        rf.sorted_active_global_handles_,
+        /* is_start */ false,
+        rf);
+    mergeRunCallbacks(
+        sorted_tls_callbacks_,
+        rf.sorted_active_tls_handles_,
+        /* is_start */ false,
+        rf);
   }
 
-  inline bool shouldRunCallback(size_t cb_idx, RecordScope scope) const {
-    TORCH_INTERNAL_ASSERT(cb_idx < callbacks_.size());
-    return callbacks_[cb_idx].scopes_[static_cast<size_t>(scope)] &&
-           (!callbacks_[cb_idx].is_sampled_ ||
-            (sample_zero_one() < samplingProbability(cb_idx)));
+ private:
+  bool tryRunCallback(const std::function<void(const RecordFunction&)>& fn, RecordFunction& rf) {
+    try {
+      fn(rf);
+      return true;
+    } catch (const std::exception &e) {
+      LOG(WARNING) << "Exception in RecordFunction callback: "
+                    << e.what();
+      return false;
+    } catch (...) {
+      LOG(WARNING) << "Exception in RecordFunction callback: unknown";
+      return false;
+    }
   }
 
-  struct Callback;
-  std::vector<Callback> callbacks_;
-
-  double global_prob_ = 0.0;
-  bool use_global_prob_ = false;
-  bool has_callbacks_with_inputs_ = false;
-
-  // tracks the current 'version' of callbacks;
-  // every time we push or pop callbacks, we bump this counter
-  uint64_t callbacks_version_ = 0;
-
-  struct Callback {
-    Callback(
-        std::function<bool(const RecordFunction&)> start_cb,
-        std::function<void(const RecordFunction&)> end_cb,
-        bool needs_inputs,
-        double sampling_prob,
-        std::unordered_set<RecordScope, std::hash<RecordScope>> scopes
-    ) : start_cb_(std::move(start_cb)),
-        end_cb_(std::move(end_cb)),
-        needs_inputs_(needs_inputs),
-        sampling_prob_(sampling_prob),
-        is_sampled_(sampling_prob != 1.0) {
-      if (!scopes.empty()) {
-        scopes_.fill(false);
-        for (auto sc : scopes) {
-          scopes_[static_cast<size_t>(sc)] = true;
+  void mergeRunCallbacks(
+      const idd_callbacks_vector& sorted_callbacks,
+      handles_vector sorted_handles,
+      bool is_start,
+      RecordFunction& rf) {
+    size_t idx_c = 0;
+    size_t idx_h = 0;
+    size_t num_executed = 0;
+    while (idx_c < sorted_callbacks.size() && idx_h < sorted_handles.size()) {
```

Comment:
nit: given that we usually expect all of the sorted_callbacks to be present, it could have been written as a for loop on sorted_handles that just bumps another index. But it's up to you